feat: Support @GQLDefault

frekw opened the pull request:
Hi!

This hopefully closes #1010.

As discussed, I've had some trouble with validating the default values. I think this approach is sort of OK, but am happy to try other paths that might lead to less code duplication.

The error messages aren't great right now because they're lacking some context. Will improve them before marking this PR as ready to go!

ghostdogpr commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,135 @@
+        }
+      case "Int"     =>
+        argValue match {
+          case BigIntNumber(value) => IO.unit
```
You can simply check if it's a `IntValue` since it's sealed.
Same below with `FloatValue`.

frekw replied:
👍

ghostdogpr commented on core/src/test/scala/caliban/execution/DefaultValueSpec.scala:
```diff
@@ -0,0 +1,195 @@
+        }
+        // NON_NULL
+      ),
+      testM("field default values") {
```
A test that shows failed validation would be nice too.

frekw replied:
👍

ghostdogpr commented on core/src/main/scala-2/caliban/parsing/Parser.scala:
```diff
@@ -1,9 +1,13 @@
 import caliban.parsing.adt._
 import fastparse._
-import zio.{ IO, Task }
+import zio.IO
```
I liked them grouped better :P

frekw replied:
Sorry, that was metals `Optimize imports`. I'll merge them again :)

frekw commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,131 @@
+                "The default value for a field must be written using GraphQL input syntax."
+              )
+            )
+        _     <- Validator.validateInputValues(field, value)
```
Not loving the circular dependency here, but maybe splitting the Validation up more is for another time.

ghostdogpr replied:
Can it make an infinite loop?

frekw replied:
No. I just have a very strong dislike towards circular dependecies. Might be Golang PTSD. 😅

ghostdogpr commented on core/src/main/scala/caliban/validation/Validator.scala:
```diff
@@ -16,7 +16,19 @@ import caliban.parsing.adt.Type.NamedType
 import caliban.parsing.adt._
 import caliban.schema.{ RootSchema, RootSchemaBuilder, RootType, Types }
 import caliban.{ InputValue, Rendering, Value }
+import caliban.parsing.Parser
```
Same here

ghostdogpr replied:
Many of them seem not used

frekw replied:
Fixed in 0f74a1d 👍

ghostdogpr commented on core/src/test/scala/caliban/execution/DefaultValueSpec.scala:
```diff
@@ -0,0 +1,203 @@
+package caliban.execution
+
+import caliban.CalibanError
```
Grouping some imports would be nice

frekw replied:
Fixed in 2c3c179 👍

ghostdogpr commented on core/src/test/scala/caliban/execution/DefaultValueSpec.scala:
```diff
@@ -0,0 +1,203 @@
+  override def spec: ZSpec[TestEnvironment, Any] =
+    suite("DefaultValueSpec")(
+      suite("default value validation")(
+        testM("string validation") {
```
Maybe mention in the test names that it's testing an invalid default value

frekw replied:
f00b649 👍

ghostdogpr commented on core/src/test/scala/caliban/execution/ExecutionSpec.scala:
```diff
@@ -9,7 +9,7 @@ import caliban.TestUtils._
 import caliban.introspection.adt.__Type
 import caliban.parsing.adt.LocationInfo
-import caliban.schema.Annotations.{ GQLInterface, GQLName, GQLValueType }
+import caliban.schema.Annotations.{ GQLDefault, GQLInterface, GQLName, GQLValueType }
```
Not needed

frekw replied:
03116cf

ghostdogpr commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,153 @@
+import caliban.Value.FloatValue.FloatNumber
+import caliban.InputValue.ObjectValue
+
+object DefaultValue {
```
`DefaultValueValidator`? To keep the naming consistent

frekw replied:
Naming is my weakness :D

945e1e4

ghostdogpr commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,153 @@
+        _     <- Validator.validateInputValues(field, value)
+        _     <- validateInputTypes(field, value, errorContext)
+      } yield ()
+    }.getOrElse(IO.unit)
```
You can do 
```scala
IO.whenCase(field.defaultValue) {
  case Some(v) => ...
}
```
to avoid this `getOrElse(IO.unit)`

frekw replied:
c55749e
Nice catch! I always seem to forget about `whenCase` 🤦 .

ghostdogpr replied:
I added it to ZIO so I know 🤣

ghostdogpr commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,153 @@
+          IO.fromEither(Parser.parseInputValue(v))
+            .mapError(e =>
+              ValidationError(
+                s"$errorContext failed to parse default value: ${e.msg}",
```
Is there always a line break at the end of the error?

frekw replied:
Hm, no I don't think so, but there are two messages, `msg` and `explanatoryText`.

ghostdogpr commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,153 @@
+              .map(_.name)
+            val exists   = possible.exists(_ == value)
+
+            if (exists) IO.unit
```
`IO.unless(exists)(fail...)`

frekw replied:
💅 👍 
afdb356

ghostdogpr reviewed (commented):
Imports management is THE reason I can't use Metals 😆 

Can you add documentation in https://github.com/ghostdogpr/caliban/blob/master/vuepress/docs/docs/schema.md#annotations ?

Looks great otherwise!

frekw commented on core/src/main/scala/caliban/validation/DefaultValue.scala:
```diff
@@ -0,0 +1,153 @@
+package caliban.validation
+
+import caliban.CalibanError.ValidationError
```
Fixed in 784667b 👍

frekw commented:
Added docs in 37dd3a1. 👍 

> Imports management is THE reason I can't use Metals 😆

I've learnt to live with it. It being automatic is just too big of a convenience 😅.  But it seems as if it hadn't run correctly on all files, so I've cleaned the imports up.

> Looks great otherwise!
🙇

ghostdogpr reviewed (approved):
Looks all good, thanks!
